torch/_functorch/aot_autograd.py

```diff
@@ -2917,8 +2922,32 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
                 torch._guards.TracingContext.get().fw_metadata = fw_metadata
 
             with TracingContext.report_output_strides() as fwd_output_strides:
+                kept_flat_args = []
+                kept_input_pos = []
+                unused_nodes = []
+                compile_bwd = True
+                for i, node in enumerate(fw_module.graph.nodes):
+                    if node.op == "placeholder":
+                        # This should be as simple as rejecting nodes without users. Unfortunately,
+                        # there are a ton of edge cases here https://github.com/pytorch/pytorch/issues/97894, activation
+                        # checkpointing, etc. So we just skip what should be a nice invariant, in favor of explicitly rejecting
+                        # functions.
+                        # if (len(node.users) == 0))
+                        if isinstance(adjusted_flat_args[i], (types.FunctionType, functools.partial)):
```

Comment:
I'm confused... why do we have function inputs to the (post-partinioning) forward graph? This should be totally unnecessary right? Because you can only ever send function hooks to the backward graph; if it could have been run in forwards, dynamo should have inlined it already.

Reply:
(Unconvinced you need all of this argument trimming business, it seems papering over the problem above.)

Reply:
iiuc we lifted the hook function to become a dynamo graph input so that we could source/guard it.  but the side effect of doing this was that it shows up in aot forward even though it's not really needed till backward.

Reply:
I think I'm still on the side of "burn hooks as attributes into the dynamo graph", so we don't have to deal with all of the extra calling convention changes and indexing work here to sort them out of the forward graph.

With the downside being that if a hook every changes, we'll need to recompile both the fw **and** bw, instead of just needing to compile the bw. But... users will rarely be dynamically changing the hooks they use from run to run, right?